### PR TITLE
Support pathlib.Path file paths when saving ONNX models

### LINKED ISSUE #19726: Incompatible LightningModule.to_onnx type-hint
### Bug description

Hello! 

The `LightningModule.to_onnx` [1] method provides a `file_path: Union[str, Path]` type-hint. However, `file_path` is later passed (unmodified) into `torch.onnx.export` [2] which expects an `Union[str, io.BytesIO]` file handle (though, I think [3] suggests it should in fact only be a `str`...)

Personally, I'd like to be able to pass `Path` instances into `LightningModule.to_onnx`, which can be achieved by changing [4] to:
```
torch.onnx.export(self, input_sample, str(file_path), **kwargs)
```
Alternatively, removing the `Path` type-hint corrects and possible confusion, and wouldn't (as far as I can see) adversely effect any existing Lightning code...

I'd be very happy to open a small PR (let me know if you'd prefer to cast `file_path` to a `str`, or simply drop the `Path` type-hint)! Many thanks

- [1] https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/core/module.py#L1357
- [2] https://github.com/pytorch/pytorch/blob/main/torch/onnx/utils.py#L193
- [3] https://github.com/pytorch/pytorch/blob/main/torch/onnx/utils.py#L1603
- [4] https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/core/module.py#L1398

### What version are you seeing the problem on?

master

### How to reproduce the bug

```python
from pathlib import Path
from lightning.pytorch.demos.boring_classes import BoringModel

model = BoringModel()

file_path = Path("./model.onnx")
model.to_onnx(file_path)
```

### PR DESCRIPTION
**Pull request by dominicgkerr**
## What does this PR do?

Update `LightningModule.to_onnx` to support `pathlib.Path` file paths (as previously type-hinted)

Fixes #19726

- [X] Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [X] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [X] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- ~Did you make sure to **update the documentation** with your changes? (if necessary)~
- [X] Did you write any **new necessary tests**? (not for typos and docs)
- [X] Did you verify new and **existing tests pass** locally with your changes?
- ~Did you list all the **breaking changes** introduced by this pull request?~
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [X] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [X] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--19727.org.readthedocs.build/en/19727/

### CODE REVIEW

File: tests/tests_pytorch/models/test_onnx.py [hunk 1]
```diff
@@ -42,6 +43,20 @@ def test_model_saves_with_input_sample(tmp_path):
     assert os.path.getsize(file_path) > 4e2
 
 
+@RunIf(onnx=True)
+def test_model_saves_with_pathlib_file_path(tmp_path):
+    """Test that ONNX model saves with pathlib.Path file_path and size is greater than 3 MB."""
+    model = BoringModel()
+    trainer = Trainer(fast_dev_run=True)
+    trainer.fit(model)
```
Comment (awaelchli) [inline tests/tests_pytorch/models/test_onnx.py:51]: Pretty sure it's not necessary to train. We can export the model directly.

Comment (awaelchli) [reply tests/tests_pytorch/models/test_onnx.py:51]:
```suggestion
```

File: tests/tests_pytorch/models/test_onnx.py [hunk 2]
```diff
@@ -42,6 +43,20 @@ def test_model_saves_with_input_sample(tmp_path):
     assert os.path.getsize(file_path) > 4e2
 
 
+@RunIf(onnx=True)
+def test_model_saves_with_pathlib_file_path(tmp_path):
```
Comment (awaelchli) [inline tests/tests_pytorch/models/test_onnx.py:47]: Alternatively, we could just modify the test above to just test both ways.

**Review by awaelchli — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

File: tests/tests_pytorch/models/test_onnx.py (hunk 1 above)
Comment (dominicgkerr) [reply tests/tests_pytorch/models/test_onnx.py]: Good spot - I've updated the test to combine other `os.path` and `pathlib.Path` cases (and removed the `.fit(...)` call at the same time): https://github.com/Lightning-AI/pytorch-lightning/pull/19727/commits/340c9fa01aa93f88f90008dff0b40384adb1bb83
